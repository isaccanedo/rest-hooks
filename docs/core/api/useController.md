---
title: useController()
---

<head>
  <title>useController() - Imperative Controls for Rest Hooks</title>
  <meta name="docsearch:pagerank" content="10"/>
</head>

```typescript
function useController(): Controller;
```

Provides access to [Controller](./Controller.md) which can be used for imperative control
over the cache. For instance [fetch](./Controller.md#fetch), [invalidate](./Controller.md#invalidate),
and [receive](./Controller.md#receive)

```tsx
import { useController } from 'rest-hooks';

function MyComponent({ id }) {
  const { fetch, invalidate, resetEntireStore } = useController();

  const handleRefresh = useCallback(
    async e => {
      await fetch(MyResource.get, { id });
    },
    [fetch, id],
  );

  const handleSuspend = useCallback(
    async e => {
      await invalidate(MyResource.get, { id });
    },
    [invalidate, id],
  );

  const handleLogout = useCallback(
    async e => {
      resetEntireStore();
    },
    [resetEntireStore],
  );
}
```
