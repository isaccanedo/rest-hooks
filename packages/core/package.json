{
  "name": "@rest-hooks/core",
  "version": "3.4.0",
  "description": "Asynchronous data framework for React",
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md",
    "./typescript.svg",
    "./rest_hooks_logo_and_text.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:js:node": "BROWSERSLIST_ENV=node12 rollup -c",
    "build:js:browser": "BROWSERSLIST_ENV=legacy rollup -c",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo ./index.d.ts",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4 && copyfiles --up 1 ./src-legacy-types/**/*.d.ts ./ts3.4/",
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "dev": "yarn run build:lib -w",
    "prepare": "yarn run build:lib",
    "prepack": "yarn prepare",
    "prepublishOnly": "yarn run build:bundle && yarn run build:legacy:lib"
  },
  "keywords": [
    "react",
    "data",
    "cache",
    "flux",
    "suspense",
    "fetch",
    "hook",
    "networking",
    "async",
    "concurrent mode",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "GraphQL",
    "query",
    "react-native",
    "ios",
    "android",
    "web",
    "middleware",
    "declarative",
    "scalable"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "license": "Apache-2.0",
  "homepage": "https://resthooks.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "@rest-hooks/normalizr": "^9.2.0",
    "@rest-hooks/use-enhanced-reducer": "^1.2.0",
    "flux-standard-action": "^2.1.1"
  },
  "peerDependencies": {
    "@types/react": "^16.8.4 || ^17.0.0 || ^18.0.0-0",
    "react": "^16.8.4 || ^17.0.0 || ^18.0.0-0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-dts": "^5.0.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  }
}
