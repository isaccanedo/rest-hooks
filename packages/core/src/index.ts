export * as __INTERNAL__ from './internal.js';
export type {
  NetworkError,
  UnknownError,
  ErrorTypes,
  Schema,
  EndpointInterface,
  EntityInterface,
  ResolveType,
} from '@rest-hooks/normalizr';
export { ExpiryStatus } from '@rest-hooks/normalizr';
export {
  default as NetworkManager,
  ResetError,
} from './state/NetworkManager.js';
export {
  default as createReducer,
  initialState,
} from './state/createReducer.js';
export { default as reducer } from './state/reducerInstance.js';
export { default as applyManager } from './state/applyManager.js';
export { useDenormalized } from './state/selectors/index.js';
export {
  useCache,
  useFetcher,
  useController,
  useFetchDispatcher,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useError,
  CacheProvider,
  BackupBoundary,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  hasUsableData,
} from './react-integration/index.js';
export {
  useSuspense,
  useFetch,
  useDLE,
} from './react-integration/newhooks/index.js';
export {
  StateContext,
  DispatchContext,
  DenormalizeCacheContext,
  ControllerContext,
  StoreContext,
  type Store,
} from './react-integration/context.js';
export { default as Controller } from './controller/Controller.js';

export * from './controller/types.js';
export * from './state/actions/index.js';
export * as actionTypes from './actionTypes.js';
export { usePromisifiedDispatch } from '@rest-hooks/use-enhanced-reducer';
/* istanbul ignore next */
export * from './types.js';
export type {
  FetchShape,
  ReadShape,
  MutateShape,
  DeleteShape,
} from './endpoint/shapes.js';
export type {
  SetShapeParams,
  ParamsFromShape,
  BodyFromShape,
  ReturnFromShape,
} from './endpoint/types.js';
