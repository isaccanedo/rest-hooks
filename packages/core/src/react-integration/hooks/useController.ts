import { useContext } from 'react';

import Controller from '../../controller/Controller.js';
import { ControllerContext } from '../context.js';

/**
 * Imperative control of Rest Hooks store
 * @see https://resthooks.io/docs/api/useController
 */
export default function useController(): Controller {
  return useContext(ControllerContext);
}
