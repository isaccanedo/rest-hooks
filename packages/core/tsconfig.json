{
  "extends": "../../tsconfig-base",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": ["src"],
  "references": [
    { "path": "../../__tests__" },
    { "path": "../normalizr" },
    { "path": "../use-enhanced-reducer" }
  ]
}
