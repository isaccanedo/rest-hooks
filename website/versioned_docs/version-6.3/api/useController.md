---
title: useController()
---

<head>
  <title>useController() - Imperative Controls for Rest Hooks</title>
</head>

```typescript
function useController(): Controller;
```

Provides access to [Controller](./Controller.md)

```tsx
import { useController } from 'rest-hooks';

function MyComponent({ id }) {
  const { fetch, invalidate, resetEntireStore } = useController();

  const handleRefresh = useCallback(
    async e => {
      await fetch(MyResource.detail(), { id });
    },
    [fetch, id],
  );

  const handleSuspend = useCallback(
    async e => {
      await invalidate(MyResource.detail(), { id });
    },
    [invalidate, id],
  );

  const handleLogout = useCallback(
    async e => {
      resetEntireStore();
    },
    [resetEntireStore],
  );
}
```
