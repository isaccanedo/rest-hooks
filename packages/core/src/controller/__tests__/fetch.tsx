import { CoolerArticle, FutureArticleResource } from '__tests__/new';
import nock from 'nock';
import { useCache, useSuspense, ResolveType } from '@rest-hooks/core';
import { act } from '@testing-library/react-hooks';

// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderRestHook,
  makeCacheProvider,
  makeExternalCacheProvider,
  FixtureEndpoint,
} from '../../../../test';
import useController from '../../react-integration/hooks/useController';

export const payload = {
  id: 5,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const createPayload = {
  id: 1,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const nested = [
  {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 3,
    title: 'the next time',
    content: 'whatever',
    author: {
      id: 23,
      username: 'charles',
      email: 'bob@bob.com',
    },
  },
];

describe.each([
  ['CacheProvider', makeCacheProvider],
  ['ExternalCacheProvider', makeExternalCacheProvider],
] as const)(`%s`, (_, makeProvider) => {
  // TODO: add nested resource test case that has multiple partials to test merge functionality

  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  let mynock: nock.Scope;

  beforeEach(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload)
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '')
      .delete(`/article/${payload.id}`)
      .reply(200, {})
      .get(`/article-cooler/0`)
      .reply(403, {})
      .get(`/article-cooler/666`)
      .reply(200, '')
      .get(`/article-cooler`)
      .reply(200, nested)
      .post(`/article-cooler`)
      .reply(200, createPayload);

    mynock = nock(/.*/).defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  beforeEach(() => {
    renderRestHook = makeRenderRestHook(makeProvider);
  });

  let errorspy: jest.SpyInstance;
  beforeEach(() => {
    errorspy = jest.spyOn(global.console, 'error');
  });
  afterEach(() => {
    errorspy.mockRestore();
  });

  it('should fetch', async () => {
    const { result } = renderRestHook(() => {
      return {
        data: useCache(FutureArticleResource.get, payload.id),
        fetch: useController().fetch,
      };
    });
    expect(result.current.data).toBeUndefined();
    let response;
    await act(async () => {
      result.current.fetch(FutureArticleResource.get, payload.id);
      result.current.fetch(FutureArticleResource.get, payload.id);
      result.current.fetch(FutureArticleResource.get, payload.id);
      response = await result.current.fetch(
        FutureArticleResource.get,
        payload.id,
      );
    });
    expect(result.current.data).toBeDefined();
    expect(result.current.data?.content).toEqual(payload.content);
    expect(response).toEqual(payload);

    // type tests
    // TODO: move these to own unit tests if/when applicable
    () => {
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.get);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.get, {
        id: payload.id,
      });
      result.current.fetch(FutureArticleResource.create, payload);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.create, {}, payload);
      result.current.fetch(FutureArticleResource.update, payload.id, payload);
      // @ts-expect-error
      result.current.fetch(FutureArticleResource.update, payload);
    };
  });

  it('should fetch with resolver', async () => {
    // we use this id because it is not nock'd
    const id = 10000;
    const fixture = {
      endpoint: FutureArticleResource.get,
      args: [10000],
      response: payload,
    };
    const { result } = renderRestHook(
      () => {
        return {
          data: useCache(FutureArticleResource.get, id),
          fetch: useController().fetch,
        };
      },
      { resolverFixtures: [fixture] },
    );
    expect(result.current.data).toBeUndefined();
    let response;
    await act(async () => {
      result.current.fetch(FutureArticleResource.get, id);
      result.current.fetch(FutureArticleResource.get, id);
      result.current.fetch(FutureArticleResource.get, id);
      response = await result.current.fetch(FutureArticleResource.get, id);
    });
  });

  it('should update on create', async () => {
    const endpoint = FutureArticleResource.create;
    const response: ResolveType<typeof endpoint> = createPayload;
    const fixtures: FixtureEndpoint[] = [
      {
        endpoint,
        args: [{ id: 1 }],
        response,
      },
    ];
    // use nock, and use resolver
    for (const resolverFixtures of [undefined, fixtures]) {
      const { result, waitForNextUpdate } = renderRestHook(
        () => {
          const articles = useSuspense(FutureArticleResource.getList);
          const fetch = useController().fetch;
          return { articles, fetch };
        },
        { resolverFixtures },
      );
      await waitForNextUpdate();
      await act(async () => {
        const article = await result.current.fetch(
          FutureArticleResource.create,
          {
            id: 1,
          },
        );
        article.content;
        // @ts-expect-error
        article.asdf;
        // @ts-expect-error
        article.pk;
      });

      expect(result.current.articles.map(({ id }) => id)).toEqual([1, 5, 3]);
    }
  });

  it('should log error message when user update method throws', async () => {
    const endpoint = FutureArticleResource.create.extend({
      update: () => {
        throw new Error('usererror');
      },
    });
    const response: ResolveType<typeof endpoint> = createPayload;
    const fixtures: FixtureEndpoint[] = [
      {
        endpoint,
        args: [{ id: 1 }],
        response,
      },
    ];
    const { result, waitForNextUpdate } = renderRestHook(
      () => {
        const articles = useSuspense(FutureArticleResource.getList);
        const fetch = useController().fetch;
        return { articles, fetch };
      },
      { resolverFixtures: fixtures },
    );
    await waitForNextUpdate();
    await act(async () => {
      await result.current.fetch(endpoint, {
        id: 1,
      });
    }),
      // still keeps old list
      expect(result.current.articles.map(({ id }) => id)).toEqual([5, 3]);

    expect(errorspy.mock.calls).toMatchSnapshot();
  });

  it('should not suspend once deleted and redirected at same time', async () => {
    const temppayload = {
      ...payload,
      id: 1234,
    };
    mynock
      .get(`/article-cooler/${temppayload.id}`)
      .reply(200, temppayload)
      .delete(`/article-cooler/${temppayload.id}`)
      .reply(204, '');
    const throws: Promise<any>[] = [];
    const { result, waitForNextUpdate, rerender } = renderRestHook(
      ({ id }) => {
        try {
          return [
            useSuspense(FutureArticleResource.get, id) ?? null,
            useController().fetch,
          ] as const;
        } catch (e: any) {
          if (typeof e.then === 'function') {
            if (e !== throws[throws.length - 1]) {
              throws.push(e);
            }
          }
          throw e;
        }
      },
      { initialProps: { id: temppayload.id as number | null } },
    );
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    const [data, fetch] = result.current;
    expect(data).toBeInstanceOf(CoolerArticle);
    expect(data?.title).toBe(temppayload.title);
    expect(throws.length).toBe(1);

    mynock
      .persist()
      .get(`/article-cooler/${temppayload.id}`)
      .reply(200, { ...temppayload, title: 'othertitle' });

    expect(throws.length).toBe(1);
    await act(async () => {
      await fetch(FutureArticleResource.delete, temppayload.id);
      rerender({ id: null });
    });
    expect(throws.length).toBe(1);
    expect(result.current[0]).toBe(null);
  });
});
