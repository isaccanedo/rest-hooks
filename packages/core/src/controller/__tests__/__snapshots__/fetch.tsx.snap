// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheProvider should log error message when user update method throws 1`] = `
[
  [
    "The following error occured during Endpoint.update() for POST http://test.com/article-cooler",
  ],
  [
    [Error: usererror],
  ],
]
`;

exports[`ExternalCacheProvider should log error message when user update method throws 1`] = `
[
  [
    "The following error occured during Endpoint.update() for POST http://test.com/article-cooler",
  ],
  [
    [Error: usererror],
  ],
]
`;
